crates/wysiwyg/src/composer_model/lists.rs

```diff
@@ -357,4 +359,464 @@ where
+            self.state.dom.lookup_node(&list_item_handle)
+        {
+            if !list_item.is_list_item() {
+                panic!("Parent node must be a list item");
+            }
+            if let DomNode::Container(list) = self
+                .state
+                .dom
+                .lookup_node_mut(&list_item_handle.parent_handle())
+            {
+                list.remove_child(list_item_handle.index_in_parent())
+            } else {
+                panic!("ListItem mus have a parent");
+            }
+        } else {
+            panic!("Parent must be a ContainerNode");
+        };
+
+        if let DomNode::Container(into_node) =
+            self.state.dom.lookup_node_mut(&into_handle)
+        {
+            // New list node added here, insert it into that container at index 0
+            if at_index < into_node.children().len() {
+                if let Some(DomNode::Container(sub_node)) =
+                    into_node.get_child_mut(at_index)
+                {
+                    sub_node.insert_child(0, removed_list_item);
+                }
+            } else if into_node.is_list_of_type(parent_list_type.clone()) {
+                into_node.insert_child(at_index, removed_list_item);
+            } else {
+                let new_list = DomNode::new_list(
+                    parent_list_type.clone(),
+                    vec![removed_list_item],
+                );
+                into_node.insert_child(at_index, new_list);
+            }
+        }
+    }
+
+    fn get_last_list_node_in_list_item(
+        &mut self,
+        handle: &DomHandle,
+        list_type: &ListType,
+    ) -> Option<&mut ContainerNode<S>> {
+        if let DomNode::Container(prev_sibling) =
+            self.state.dom.lookup_node_mut(&handle)
+        {
+            if !prev_sibling.children().is_empty() {
+                if let DomNode::Container(prev_sibling_last_item) = prev_sibling
+                    .get_child_mut(prev_sibling.children().len() - 1)
+                    .unwrap()
+                {
+                    if prev_sibling_last_item.is_list_of_type(list_type.clone())
+                    {
+                        return Some(prev_sibling_last_item);
+                    }
+                }
+            }
+        }
+        None
+    }
+
+    pub fn un_indent_locations(&mut self, locations: Vec<DomLocation>) {
+        self.un_indent_handles(Self::leaf_handles_from_locations(locations));
+    }
+
+    fn un_indent_handles(&mut self, handles: Vec<DomHandle>) {
+        let by_list_sorted = Self::group_sorted_handles_by_list_parent(handles);
+
+        for (list_handle, handles) in by_list_sorted.iter().rev() {
+            let mut sorted_handles = handles.clone();
+            sorted_handles.sort();
+
+            let at_index = list_handle.parent_handle().index_in_parent() + 1;
+            if let Some(into_handle) =
+                self.find_closest_list_ancestor(list_handle)
```

Comment:
Somewhere near here the compile failed because I changed some interfaces to take `&DomHandle`, sorry!

Reply:
Will rebase and fix it 👍